shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -560,6 +591,15 @@ private void ensureAttachedToNative() {
     }
   }
 
+  private void ensureRunningOnMainThread() {
+    if (BuildConfig.DEBUG && Looper.myLooper() != Looper.getMainLooper()) {
+      throw new RuntimeException(
+          "Attempted to invoke a @UiThread method on a different thread ("
+              + Thread.currentThread().getName() + ")"
+      );
+    }
```

Comment:
Worth an error log in release builds?

Reply:
I'm not clear on why we'd guard this as debug only - it will crash in both release and debug, right?

Maybe just do a `Log.f` here for all modes?

Reply:
hah, I guess it's actually `Log.wtf`

Reply:
@jason-simmons recommended throwing in debug mode. The bug that lead to this change seemed to suggest that failures were intermittent, meaning some of these calls can survive under certain conditions. If a developer has already shipped a broken app, do we want to force 100% crashes in that environment?

Reply:
These methods should never be called outside the main thread.  I was hoping to avoid the cost of the thread check in release mode, but if that cost is trivial then we can always do it.

Reply:
Am I misunderstanding that this call would _always_ result in a crash under this condition?  It seems more likely that the intermittent nature is because of some other code path leading to this call rather than that this call somehow sometimes doesn't crash.

Reply:
I don't know the answer to that.

Reply:
Invoking these APIs from a thread other than the main thread is a race condition that will yield unpredictable behavior.

Reply:
It should be pretty cheap if we cache the result of `Looper.getMainLooper()` - as long as that's safe to do (I think it is)

Reply:
getMainLooper requires a synchronization: https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/os/Looper.java#L133